src/components/UpgradeButton.js

```diff
@@ -1,22 +1,63 @@
-import React from "react";
+import React, { useState } from "react";
 import PropTypes from "prop-types";
 
-const UpgradeButton = ({ type }) => {
+import Button from "@material-ui/core/Button";
+import Menu from "@material-ui/core/Menu";
+import MenuItem from "@material-ui/core/MenuItem";
+
+const UpgradeButton = ({
+  type,
+  upgrades,
+  upgradeFunction,
+  unitKey,
+  upgradeKey
+}) => {
+  const [anchorEl, setAnchorEl] = useState(null);
+  const [selectedUpgrade, setSelectedUpgrade] = useState(undefined);
+
+  const handleClick = event => {
+    setAnchorEl(event.currentTarget);
+  };
+
+  const handleClose = upgrade => {
+    setSelectedUpgrade(upgrade);
+    upgradeFunction(unitKey, upgradeKey, upgrade);
+    setAnchorEl(null);
+  };
+
   return (
-    <button
-      onClick={e => {
-        // e.preventDefault();
-        e.stopPropagation();
-        alert(`todo: Add a ${type} upgrade`);
-      }}
-    >
-      {type}
-    </button>
+    <>
+      <Button variant="contained" onClick={handleClick}>
+        {selectedUpgrade ? selectedUpgrade.name : type}
+      </Button>
+      <Menu
+        id="upgrade-menu"
+        anchorEl={anchorEl}
+        keepMounted
+        open={Boolean(anchorEl)}
```

Comment:
```suggestion
        open={anchorEl != null}
```

Reply:
Why is this better?

Reply:
Just readability: "turn the anchorEl into a boolean" vs "is anchorEl null" - up to you 😎

Reply:
Gotcha thanks!

Reply:
`null == undefined && null == null`, but `null !== undefined`. Using `!= null` catches both undefined and null!

Reply:
JavaScript is so dumb

Reply:
Hehehe